File: behavioral/template-method/template_method.py

# Abstract Class
class AbstractClass:
    def template_method(self):
        self.primitive_operation1()
        self.primitive_operation2()

    def primitive_operation1(self):
        pass

    def primitive_operation2(self):
        pass

# Concrete Class
class ConcreteClass(AbstractClass):
    def primitive_operation1(self):
        print("ConcreteClass.primitive_operation1")

    def primitive_operation2(self):
        print("ConcreteClass.primitive_operation2")

if __name__ == "__main__":
    instance = ConcreteClass()
    instance.template_method()

File: src/behavioral/template-method/TemplateMethod.cs
namespace DesignPatterns.Behavioral.TemplateMethod;

// Abstract Class
public abstract class AbstractClass
{
    public void TemplateMethod()
    {
        PrimitiveOperation1();
        PrimitiveOperation2();
    }

    public abstract void PrimitiveOperation1();
    public abstract void PrimitiveOperation2();
}

// Concrete Class
public class ConcreteClass : AbstractClass
{
    public override void PrimitiveOperation1()
    {
        Console.WriteLine("ConcreteClass.PrimitiveOperation1");
    }

    public override void PrimitiveOperation2()
    {
        Console.WriteLine("ConcreteClass.PrimitiveOperation2");
    }
}

class Program
{
    static void Main()
    {
        AbstractClass instance = new ConcreteClass();
        instance.TemplateMethod();
    }
}
